### PR TITLE
Extend the LEP to allow customization of the cipher

### PR DESCRIPTION
**Pull request by mantissahz**
Extend the old LEP `pv-encryption.md` to allow users customize the options of cipher used by `cryptsetup`.

longhorn/longhorn#3353

Signed-off-by: James Lu <james.lu@suse.com>

### CODE REVIEW

File: enhancements/20221024-pv-encryption.md [hunk 1]
```diff
@@ -29,20 +29,20 @@ this in return means that backups of that volume end up also being encrypted.
...
 ### User Stories
-All regular longhorn operations should also be supported for encrypted volumes, 
+All regular longhorn operations should also be supported for encrypted volumes,
 therefore the only user story that is mentioned is
 how to create and use an encrypted volume.
 
 #### Create and use an encrypted volume
 - create a storage class with (encrypted=true) and either a global secret or a per volume secret
-- create the secret for that volume in the configured namespace
+- create the secret for that volume in the configured namespace with customization options of the cipher for instance `cipher`, `key-size` and `hash`
```
Comment (c3y1huang) [inline enhancements/20221024-pv-encryption.md:38]: Do we need to add those to the test-plan?

File: enhancements/20221024-pv-encryption.md [hunk 2]
```diff
@@ -112,18 +115,26 @@ parameters:
...
+- `CRYPTO_KEY_CIPHER` allow users to choose the cipher algorithm when creating an encrypted volume by `cryptsetup`,
+- `CRYPTO_KEY_HASH` specifies the hash used in the LUKS key setup scheme and volume key digest
+- `CRYPTO_KEY_SIZE` sets key size in bits. The argument has to be a multiple of 8 and maximum interactive passphrase length is 512 (characters.)
+
 ```yaml
   CRYPTO_KEY_VALUE: "Simple passphrase"
   CRYPTO_KEY_PROVIDER: "secret" # this is optional we currently only support direct keys via secrets
+  CRYPTO_KEY_CIPHER: "aes-xts-plain64" # this is optional
+  CRYPTO_KEY_HASH: "sha256" # this is optional
+  CRYPTO_KEY_SIZE: "256" # this is optional
```
Comment (derekbit) [inline enhancements/20221024-pv-encryption.md:137]: What are the default values? Any tradeoff about different settings?

Comment (mantissahz) [reply enhancements/20221024-pv-encryption.md:137]: The default values are listed as the example.
Yes, cipher and key-size have the dependency, some ciphers could not use 512 key-size.
And There are different speeds of encryption and decryption for different cipher choices.

File: enhancements/20221024-pv-encryption.md [hunk 3]
```diff
@@ -112,18 +115,26 @@ parameters:
...
+Storing the `Encrypted` value is necessary to support encryption for RWX volumes.
 
 ## Design
 
 ### Implementation Overview
 Host requires `dm_crypt` kernel module as well as `cryptsetup` installed.
-We utilize the below parameters from a secret, `CRYPTO_KEY_PROVIDER` allows us in the future to add other key management systems.
+We utilize the below parameters from a secret,
+- `CRYPTO_KEY_PROVIDER` allows us in the future to add other key management systems,
+- `CRYPTO_KEY_CIPHER` allow users to choose the cipher algorithm when creating an encrypted volume by `cryptsetup`,
```
Comment (c3y1huang) [inline enhancements/20221024-pv-encryption.md:128]:
```suggestion
- `CRYPTO_KEY_CIPHER` allow users to choose the cipher algorithm when creating an encrypted volume by `cryptsetup`
```

File: enhancements/20221024-pv-encryption.md [hunk 4]
```diff
@@ -112,18 +115,26 @@ parameters:
...
+- `CRYPTO_KEY_HASH` specifies the hash used in the LUKS key setup scheme and volume key digest
+- `CRYPTO_KEY_SIZE` sets key size in bits. The argument has to be a multiple of 8 and maximum interactive passphrase length is 512 (characters.)
```
Comment (c3y1huang) [inline enhancements/20221024-pv-encryption.md:130]:
```suggestion
- `CRYPTO_KEY_SIZE` sets the key size in bits. The argument has to be a multiple of 8 and the maximum interactive passphrase length is 512 (characters)
```

File: enhancements/20221024-pv-encryption.md [hunk 5]
```diff
@@ -146,6 +157,14 @@ We utilize the below parameters from a secret, `CRYPTO_KEY_PROVIDER` allows us i
 - create a pod that uses that pvc for a volume mount
 - wait for pod up and healthy
 
+#### Successful Creation of an encrypted volume with customization of the cipher
+- create a storage class with (encrypted=true) and either a global secret or a per volume secret
+- create the secret with customized options of the cipher for that volume in the configured namespace
+- create a pvc that references the created storage class
+- create a pod that uses that pvc for a volume mount
+- wait for pod up and healthy
+- check if customized options of the cipher are correct
```
Comment (c3y1huang) [inline enhancements/20221024-pv-encryption.md:166]:
```suggestion
- check if the customized options of the cipher are correct
```

File: enhancements/20221024-pv-encryption.md [hunk 6]
```diff
@@ -16,7 +16,7 @@ this in return means that backups of that volume end up also being encrypted.
 ## Motivation
 
 ### Goals
-- user is able to create & use an encrypted volume
+- user is able to create & use an encrypted volume with customization options of the cipher
```
Comment (c3y1huang) [inline enhancements/20221024-pv-encryption.md:19]:
```suggestion
- user is able to create & use an encrypted volume with cipher customization options
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: enhancements/20221024-pv-encryption.md (hunk 3 above)
Comment (mantissahz) [reply enhancements/20221024-pv-encryption.md]: Modified. Thanks.

File: enhancements/20221024-pv-encryption.md (hunk 4 above)
Comment (mantissahz) [reply enhancements/20221024-pv-encryption.md]: Modified. Thanks.

File: enhancements/20221024-pv-encryption.md (hunk 5 above)
Comment (mantissahz) [reply enhancements/20221024-pv-encryption.md]: Modified. Thanks.

File: enhancements/20221024-pv-encryption.md (hunk 6 above)
Comment (mantissahz) [reply enhancements/20221024-pv-encryption.md]: Modified. Thanks.

File: enhancements/20221024-pv-encryption.md (hunk 2 above)
Comment (derekbit) [reply enhancements/20221024-pv-encryption.md:137]: Is CRYPTO_KEY_SIZE  necessary If `CRYPTO_KEY_HASH` is known?

Comment (mantissahz) [reply enhancements/20221024-pv-encryption.md:137]: No, it seems that they are not dependent on each other.

**Review by joshimoo — APPROVED** [review]
LGTM

**Review by c3y1huang — APPROVED** [review]

**Review by innobead — APPROVED** [review]
LGTM.

But think further, probably we need to have specific settings for each provider in the future, even though right now all default to dm_crypt except CRYPTO_KEY_PROVIDER.

**Pull request merged**